src/main/java/edu/harvard/iq/dataverse/pidproviders/doi/crossref/CrossRefDOIRegisterService.java

```diff
@@ -0,0 +1,326 @@
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import edu.harvard.iq.dataverse.*;
+import edu.harvard.iq.dataverse.pidproviders.AbstractPidProvider;
+import jakarta.enterprise.inject.spi.CDI;
+import org.apache.commons.text.StringEscapeUtils;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static edu.harvard.iq.dataverse.util.SystemConfig.getDataverseSiteUrlStatic;
+
+public class CrossRefDOIRegisterService implements java.io.Serializable {
+    private static final Logger logger = Logger.getLogger(CrossRefDOIRegisterService.class.getCanonicalName());
+
+    private final String url;
+    private final String apiUrl;
+    private final String username;
+    private final String password;
+    private final String depositor;
+    private final String depositorEmail;
+
+
+    private static DataverseServiceBean dataverseService = null;
+
+    private CrossRefRESTfullClient client = null;
+
+    public CrossRefDOIRegisterService(String url, String apiUrl, String username, String password, String depositor, String depositorEmail) {
+        this.url = url;
+        this.apiUrl = apiUrl;
+        this.username = username;
+        this.password = password;
+        this.depositor = depositor;
+        this.depositorEmail = depositorEmail;
+    }
+
+    private CrossRefRESTfullClient getClient() {
+        if (client == null) {
+            client = new CrossRefRESTfullClient(url, apiUrl, username, password);
+        }
+        return client;
+    }
+
+    public boolean testDOIExists(String identifier) {
+        boolean doiExists;
+        try {
+            CrossRefRESTfullClient client = getClient();
+            doiExists = client.testDOIExists(identifier.substring(identifier.indexOf(":") + 1));
+        } catch (Exception e) {
+            logger.log(Level.INFO, identifier, e);
+            return false;
+        }
+        return doiExists;
+    }
+
+    public HashMap<String, String> getMetadata(String identifier) throws IOException {
+        HashMap<String, String> metadata = new HashMap<>();
+        try {
+            CrossRefRESTfullClient client = getClient();
+            String jsonMetadata = client.getMetadata(identifier.substring(identifier.indexOf(":") + 1));
+            Map<String, Object> mappedJson = new ObjectMapper().readValue(jsonMetadata, HashMap.class);
+            logger.log(Level.FINE, jsonMetadata);
+            metadata.put("_status", mappedJson.get("status").toString());
+        } catch (RuntimeException e) {
+            logger.log(Level.INFO, identifier, e);
+        }
+        return metadata;
+    }
+
+    public String reserveIdentifier(String identifier, DvObject dvObject) throws IOException {
+        logger.info("Crossref reserveIdentifier");
```

Comment:
and probably fine for all of these that just tell you that you used the method

Reply:
changed